### DISCUSSION TITLE
OSError: [Errno 5] EIO when accessing sensor data

### DISCUSSION THREAD

**Discussion by adysanro** [micropython/micropython#9201] [Hardware & Peripherals] [answered]
Hi, 
I am being plagued by the message "OSError: [Errno 5] EIO" when accessing sensor data. Working with a BME280 sensor and a Raspberry Pi Pico W. I was able to read sensor data consistently, without an issue, up until a few days ago. But then "something" happened.
I see a lot of discussions online about this error message, many say it's related to bad soldering. I redid the soldering on all of the sensor's pins but still the same message.

Behavior: sometimes data is read for a few moments then the error is shown, most of the time I just get the error directly.
Restarting Thonny and reconnecting the board doesn't seem to make it better. When I get the error message the sensor is afterwards still found by i2c.scan(). Actually both sensors, because I have two of them connected in series as shown here: https://www.hackster.io/mr-alam/how-to-use-i2c-pins-in-raspberry-pi-pico-i2c-scanner-code-8f489f . 
I don't have any issues reading data from the other sensor, a SGP40.
The same thing happens with the BME sensor if I connect it directly to the board and disconnect the SGP sensor.

![2022-09-03 14_30_44-](https://user-images.githubusercontent.com/87180477/188268426-172c5fef-9ecf-4851-bfc7-d72a141fb32c.png)
![2022-09-03 14_42_28-Thonny  -  Raspberry Pi Pico __ _code-demo-orig py  @  29 _ 1](https://user-images.githubusercontent.com/87180477/188268850-502f7af0-9087-4435-8f70-af9de3880fd2.png)
![302161329_431607188947496_8759186945881328366_n](https://user-images.githubusercontent.com/87180477/188269047-26c14f39-d04c-482c-b66b-62d7813571a0.jpg)

[code.zip](https://github.com/micropython/micropython/files/9482461/code.zip)

Any help would be appreciated.

Thanks,
Adrian

**Comment by robert-hh** [comment]
EIO5 is communication error. Could be wiring, missing pull-up resistors, wrong address. 
Does the bme280 work if the. other sensor is disconnected? 
Are the other pins on the bme280 breakout internally open or floating?

**Reply by adysanro** [reply]
BME280 sensor still works intermittently for a few moments if connected alone. Don't know about 'internal open or floating'. I am new to this and don't know where to look...

**Reply by adysanro** [reply]
Now I understand what you meant by internally open or floating. This is what was causing the issue. The hardware address was always fluctuating between 0x76 and 0x77. After putting a wire from SDO to GND now it's always 0x76 and I'm getting data consistently without errors.

**Comment by jimmo** [comment, marked as answer]
I don't see any pull-up resistors on either of those sensor breakouts. Try adding 10k resistors from both SDA and SCL to Vcc.

**Reply by jimmo** [reply]
Looking at the hackster.io post you linked to, the don't mention pull-ups (they really should!!) however the renderings of the boards they're using all appear to plausibly have pull-ups on them. Your breakouts are different and do not appear to.

**Reply by adysanro** [reply]
In the image below from the hackster.io post I don't see any resistors. Or you are reffering to the fact that it might be built-in into the sensor? As well, the sensor works in the same way, intermittently, if connected indepedently to the board and with the other sensor disconnected.  Also, when it used to work previously a few days ago, it worked for a least 5 day almost 24/7 without errors.
![Raspberry-Pi-Pico-I2C-Example](https://user-images.githubusercontent.com/87180477/188270396-f40c522d-0ae1-45d5-aaca-a94ce4bbe1c5.jpg)

**Reply by jimmo** [reply]
> Or you are reffering to the fact that it might be built-in into the sensor?

Yes. I can see some SMD resistors on those boards (whether they're pull-ups or not I cannot tell) whereas on your sensor boards there are no resistors at all.

>  As well, the sensor works in the same way, intermittently, if connected indepedently to the board and with the other sensor disconnected.

Pull-ups would definitely be more important with more devices on the bus.

> Also, when it used to work previously a few days ago, it worked for a least 5 day almost 24/7 without errors.

Yeah I can't explain this. But given that pull-ups are mandatory for i2c to function and there's no evidence of any pull-ups, you should definitely eliminate that as the cause first.

**Reply by adysanro** [reply]
Thanks, I'll look into getting those.

**Reply by Carglglz** [reply]
@adysanro this happened to me with the exact same BME280 sensor breakout board and @robert-hh and @jimmo are right.

> I don't see any pull-up resistors on either of those sensor breakouts. Try adding 10k resistors from both SDA and SCL to Vcc.

I tried both 10k and 1k and 1k did the trick 

> Are the other pins on the bme280 breakout internally open or floating?

This is the main reason that explains this odd behaviour. There is a **SDO** Pin in the BME breakout board that is used to set the sensor I2C address. This is done by connecting it to **GND** or **VCC** (i.e  pin is pulled high). I suggest connecting it to **GND**  , after that it should be stable.

For reference the [bme280 data sheet](https://pdf1.alldatasheet.es/datasheet-pdf/view/1132060/BOSCH/BME280.html)

> The 7-bit device address is 111011x. The 6 MSB bits are fixed. The last bit is changeable by SDO value and can be changed during operation. Connecting SDO to GND results in slave address 1110110 (0x76); connection it to VDDIO results in slave address 1110111 (0x77), which is the same as BMP280’s I²C address. The SDO pin cannot be left floating; if left floating, the I²C address will be undefined.

**Reply by adysanro** [reply]
Thank you @Carglglz ! The issue is now solved 👯‍♂️ 
Now I understand what was meant by internally open or floating. This is what was causing the issue. The hardware address was always fluctuating between 0x76 and 0x77 when I was running the ic2.scan multiple times. After putting a wire from SDO to GND now it's always 0x76 and I'm getting data consistently without errors. This is with or without 10K or 4.7K resistors which I was able to get in the meantime.
So I am now thinking that both sensors have built-in resistors.

The final question. I now know that for the I2C standard you must use resistors. If the sensors are working fine without them (because maytbe they're built-in), should I still use the 10K resistors? or possible get and use 1K? I am risking to break the sensors or board long-term if not using the resistors for this specific use case?

**Reply by robert-hh** [reply]
The total pull-up resistor value should be safe down to 330 Ohm. So adding 10k in parallel to maybe existing pull-up resistors will not harm.

**Reply by adysanro** [reply]
Ok, thank you.
